src/stash/src/lib.rs

```diff
@@ -365,3 +499,89 @@ impl<K, V> From<Id> for StashCollection<K, V> {
         }
     }
 }
+
+/// A helper struct to prevent mistyping of a [`StashCollection`]'s name and
+/// k,v types.
+pub struct TypedCollection<K, V> {
+    name: &'static str,
+    typ: PhantomData<(K, V)>,
+}
+
+impl<K, V> TypedCollection<K, V> {
+    pub const fn new(name: &'static str) -> Self {
+        Self {
+            name,
+            typ: PhantomData,
+        }
+    }
+}
+
+impl<K, V> TypedCollection<K, V>
+where
+    K: Codec + Ord,
+    V: Codec + Ord,
+{
+    pub fn get(&self, stash: &impl Stash) -> Result<StashCollection<K, V>, StashError> {
+        stash.collection(self.name)
+    }
+
+    pub fn peek_one<S>(&self, stash: &S) -> Result<BTreeMap<K, V>, StashError>
+    where
+        S: Stash,
+        K: Hash,
+    {
+        stash.peek_one(self.get(stash)?)
+    }
+
+    pub fn peek_key_one(&self, stash: &impl Stash, key: &K) -> Result<Option<V>, StashError> {
+        stash.peek_key_one(self.get(stash)?, key)
+    }
+
+    /// Sets the given k,v pair.
+    pub fn upsert_key<S>(&self, stash: &S, key: &K, value: &V) -> Result<(), StashError>
+    where
+        S: Append,
+    {
+        let collection = self.get(stash)?;
+        let mut batch = collection.make_batch(stash)?;
+        let prev = match stash.peek_key_one(self.get(stash)?, key) {
+            Ok(prev) => prev,
+            Err(err) => match err.inner {
+                InternalStashError::PeekSinceUpper(_) => {
+                    // If the upper isn't > since, bump the upper and try again to find a sealed
+                    // entry. Do this by appending the empty batch which will advance the upper.
+                    stash.append(once(batch))?;
+                    batch = collection.make_batch(stash)?;
+                    stash.peek_key_one(self.get(stash)?, key)?
+                }
```

Comment:
Initially they are equal (both MIN). This specific edge case was very annoying and caused lots of bugs requiring awkward workarounds. It's not obvious to me what the better solution was beyond this.

Reply:
Would it make sense to just append an empty batch to every collection right when you create it?

Reply:
Maybe! One argument against that is the user of the Stash then has less control over the upper timestamps, because the first one is auto closed. We do have an existing use of stash (timestamp bindings) that explicitly controls all timestamps, so the collection doing something on its own like that may break some assumptions by the caller.

Reply:
Oh, that makes sense. I'm way out of my element with this stuff, so if this has already been discussed at length then feel free to ignore, but it just seems odd to me that 
1. Every collection starts with `upper == since`
2. The only way for a collection to have `(upper <= since)` is when it's first created and `upper == since`.
3. `(upper <= since)` is considered an invalid/error state.
So every collection is created in an invalid state, remains invalid until something is inserted, and then there's no way back into that invalid state.

It almost seems like `upper == since` should be a valid but empty state and only `upper < since` should be considered an error.

Reply:
It's not invalid for `upper <= since`, but it does mean the collection doesn't have any peekable entries. MZ itself can also be in this situation, but it's fine because the peek will wait for the chosen timestamp's upper to progress past the peek time. Stash doesn't have a mechanism to wait, so we have to error in peek instead.

Reply:
Maybe `peek_timestamp` would return `Timestamp::MAX` when `upper == since` so we know that no elements are returned for a collection where `upper == since`.

Reply:
Ah ok, I think the issue is that I don't fully understand the difference between "not having any peekable entries" and being empty. Though it seems like you have a much better grasp on that than I do. Thanks for the explanation!

Reply:
Yeah, if the timestamps are equal then in means we don't know what the entries will be, not that it has no entries. That's why the code has to retry the peek after bumping the upper instead of returning empty.